src/storage-client/src/controller.rs

```diff
@@ -2405,6 +2439,110 @@ where
+
+        let mut push_frontier = |object_id: GlobalId, time: T| {
+            let row = Row::pack_slice(&[
+                Datum::String(&object_id.to_string()),
+                Datum::MzTimestamp(mz_repr::Timestamp::from_bytes(time.encode())),
+            ]);
+            updates.push((row.clone(), 1));
+            old_rows.push(row);
+        };
+
+        // Add the frontiers known to the storage controller.
+        for (collection_id, collection) in &self.state.collections {
+            // Some of the storage frontiers are not advanced by us but by our users, which might
+            // provide them to us in `frontiers` too. Ensure we don't end up duplicates because of
+            // this.
+            frontiers.remove(collection_id);
+            if let Some(time) = collection.write_frontier.clone().into_option() {
+                push_frontier(*collection_id, time);
+            }
+        }
+        for (export_id, export) in &self.state.exports {
+            // Some of the storage frontiers are not advanced by us but by our users, which might
+            // provide them to us in `frontiers` too. Ensure we don't end up duplicates because of
+            // this.
+            frontiers.remove(export_id);
+            if let Some(time) = export.write_frontier.clone().into_option() {
+                push_frontier(*export_id, time);
+            }
+        }
+
+        // Add the provided frontiers.
+        for (object_id, time) in frontiers {
+            push_frontier(object_id, time);
+        }
+
+        // Cancel out superfluous retractions.
+        consolidate(&mut updates);
+
+        self.append_to_managed_collection(
+            self.state.introspection_ids[&IntrospectionType::Frontiers],
+            updates,
+        )
+        .await;
+    }
+
+    async fn record_replica_frontiers(
+        &mut self,
+        frontiers: BTreeMap<ReplicaId, Vec<(GlobalId, Self::Timestamp)>>,
+    ) {
+        let old_rows = &mut self.replica_frontiers_rows;
+        let mut updates = Vec::with_capacity(old_rows.len());
+
+        // Retract existing rows.
+        updates.extend(old_rows.drain(..).map(|row| (row, -1)));
+
+        let mut push_frontier = |replica_id, object_id: GlobalId, time: T| {
+            // TODO(#18377): Make replica IDs `NewReplicaId`s throughout the code.
+            let replica_id = NewReplicaId::User(replica_id);
+            let row = Row::pack_slice(&[
+                Datum::String(&replica_id.to_string()),
+                Datum::String(&object_id.to_string()),
+                Datum::MzTimestamp(mz_repr::Timestamp::from_bytes(time.encode())),
+            ]);
+            updates.push((row.clone(), 1));
+            old_rows.push(row);
+        };
+
+        // Add the frontiers known to the storage controller.
+        for (collection_id, collection) in &self.state.collections {
+            let Some(cluster_id) = collection.cluster_id() else { continue };
```

Comment:
Why are these collections ignored? Could we make the column nullable instead?

Reply:
These are collections that are not installed on a replica, so it makes no sense for them to appear in `mz_cluster_replica_frontiers`. I'm adding `mz_frontiers` specifically because we currently don't have any place where we expose frontiers of collections that are not maintained by replicas.

Reply:
I nominate `mz_object_frontiers` for `mz_name_tbd`

Reply:
I like that idea! A wrinkle would be that users would have to do more typing to get the global frontiers, but we could easily solve that with a view. I also don't see any dependencies on `mz_cluster_replica_frontiers` anywhere in our code, so it should be easy to fully replace it with a more general `mz_frontiers`.

I'm not entirely sure whether having just a single table would be more or less confusing to users though. Might be a question for @MaterializeInc/surfaces.

Reply:
> A wrinkle would be that users would have to do more typing to get the global frontiers, but we could easily solve that with a view. 

I didn't follow that. In my proposal both `mz_cluster_replica_frontiers` and `mz_frontiers` would be accessible with the same data as this PR is doing, just defined as views over a lower level relation (`mz_object_frontiers`). in which case would users do more typing?

Reply:
Specifically I'm thinking of making `mz_cluster_replica_frontiers` and `mz_frontiers` a `BuiltinView`. If you search for that you'll find more internal objects defined like that

Reply:
Ah, I didn't see that you already suggested adding views, sorry!

If we have the generic source, I don't think we need to create a `mz_cluster_replica_frontiers` view. Doesn't seem worth the catalog noise just for removing NULL values.

For the names, I'm not so sure about `mz_frontiers` and `mz_object_frontiers`. Seems easy to get confused about which is which. How about `mz_frontiers` and `mz_global_frontiers`?

Reply:
> I don't think we need to create a mz_cluster_replica_frontiers view. Doesn't seem worth the catalog noise just for removing NULL values.

Excellent! I agree, I thought that was needed for reasons I was not aware of.

>  How about `mz_frontiers` and `mz_global_frontiers`?

I like it! Probably worth passing it through the surfaces channel for light bikeshedding in case there is a better proposal.

Reply:
[Slack discussion.](https://materializeinc.slack.com/archives/C02FWJ94HME/p1690203737030619)